### PR TITLE
[DAG] narrowExtractedVectorBinOp - ensure we limit late node creation to LegalOperations only

### PR DESCRIPTION
**Pull request by RKSimon**
Avoids infinite issues in some upcoming patches to help D152928

### CODE REVIEW

**Comment by davemgreen** [comment]
It looks like there might be a costmodel test that needs updating? It includes some codegen tests it seems.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -23960,7 +23960,8 @@ static SDValue narrowExtractedVectorBinOp(SDNode *Extract, SelectionDAG &DAG,
   // Bail out if the target does not support a narrower version of the binop.
   EVT NarrowBVT = EVT::getVectorVT(*DAG.getContext(), WideBVT.getScalarType(),
                                    WideNumElts / NarrowingRatio);
-  if (!TLI.isOperationLegalOrCustomOrPromote(BOpcode, NarrowBVT))
+  if (!TLI.isOperationLegalOrCustomOrPromote(BOpcode, NarrowBVT,
+                                             LegalOperations))
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:23964]: This part seems obvious

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 2]
```diff
@@ -41371,8 +41371,18 @@ bool X86TargetLowering::SimplifyDemandedVectorEltsForTargetNode(
     case X86ISD::UNPCKH:
     case X86ISD::BLENDI:
       // Integer ops.
+    case X86ISD::PCMPEQ:
```
Comment (arsenm) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:41374]: Don't know enough about x86 but seems like there would need to be more tests hitting these cases?

**Comment by RKSimon** [comment]
Sorry this was my first attempt at a patch series and I screwed it up - converting to draft until I've fixed it up

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 2 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:41374]: I split this into 2 commits so the review can see the x86 regressions and the followup that addresses them - I intend to squash+merge these - do you think that's enough?

File: llvm/test/CodeGen/X86/kshift.ll [hunk 3]
```diff
@@ -270,10 +270,11 @@ define i64 @kshiftl_v64i1_63(<64 x i8> %x, <64 x i8> %y) {
 ; KNL-NEXT:    vpcmpeqb %xmm2, %xmm0, %xmm0
 ; KNL-NEXT:    vpmovsxbd %xmm0, %zmm0
 ; KNL-NEXT:    vptestmd %zmm0, %zmm0, %k0
+; KNL-NEXT:    vpxor %xmm0, %xmm0, %xmm0
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/kshift.ll:273]: This is an known existing (very old) issue we're having in that the 128/256/512-bit zero vectors don't get merged properly: #26392

**Comment by davemgreen** [comment]
I can't speak for the X86/kshift change but this looks OK to me.

**Review by davemgreen — APPROVED** [review]

**Pull request merged**